Heavy Adornment Memory Optimizations

MaximumADHD opened the pull request:
Theres some memory consumption problem with the Explorer, that happens when multiple objects are being created and destroyed during each frame. This has been a problem for several months now, but it has become more notable with this plugin, due to the CoreGui suddenly becoming visible in the Explorer.

I updated parts of the Adornment module to utilize a smart object pooling system, so that objects don't have to get cleared and regenerated during each frame. As a side effect, this actually makes the tool faster than it was before, and it can handle large parts without the framerate panicking too much.

Commit:
Heavy Adornment Memory Optimizations

Theres some memory consumption problem with the Explorer, that happens when multiple objects are being created and destroyed during each frame. This has been a problem for several months now, but it has become more notable with this plugin, due to the CoreGui suddenly becoming visible in the Explorer.

I updated parts of the Adornment module to utilize a smart object pooling system, so that objects don't have to get cleared and regenerated during each frame. As a side effect, this actually makes the tool faster than it was before, and it can handle large parts without the framerate panicking too much.

DarraghGriffin commented on TransformDragger.rbxmx:
```diff
@@ -6624,5 +6616,88 @@ return module
+spawn(function ()
+	while wait(1) do
+		for _,queue in pairs(bin) do
+			for object,expireTime in pairs(queue) do
```
This isn't too much of an issue but it seems like iterating through the entire queue table every second could be avoided by maintaining the queue in a different way. If you use a queue data structure with fast addition and removal from either end of the queue then you can just check if the top of the queue has an expiration time less than the current tick.

MaximumADHD replied:
Good call, I'll update it.

Commit:
Addressing Feedback

Objects that are recycled will individually handle their own garbage collection instead of running on a while loop.

CloneTrooper1019 commented:
Bump.

tberg-roblox commented:
Sorry, been super busy. I'll try to take a look at this today.

tberg-roblox commented on TransformDragger.rbxmx:
```diff
@@ -5768,7 +5765,7 @@ local function updatePlanePosition()
 	--if not workplaneFrame then return end
 	
-	--drawTemporaryLine(Vector3.new(0,0,0), Vector3.new(10, 10, 0), Color3.new(.3,.3,.3), 0.0, 1.5, planeKeeper[1].Adornee)
+	--appendTemporaryLine(Vector3.new(0,0,0), Vector3.new(10, 10, 0), Color3.new(.3,.3,.3), 0.0, 1.5, planeKeeper[1].Adornee)
```
These comments can be removed

MaximumADHD replied:
Got it.

tberg-roblox commented on TransformDragger.rbxmx:
```diff
@@ -6624,5 +6616,88 @@ return module
+-- Allocate works similarly to Pull, except it tries to reuse objects that exist in the specified bin.
+-- This removes the need to recycle every object during every frame.
+
+function recyclingBin:Allocate(objectType,needed,bin)
```
Please rename bin. It's confusing that we have two objects named the same thing, one containing a table, while the other contains an Instance.

MaximumADHD replied:
Whoops, small oversight.
I updated it to be called allocationStorage instead of bin.

tberg-roblox commented on TransformDragger.rbxmx:
```diff
@@ -6624,5 +6616,88 @@ return module
+	if not bin[class] then
+		bin[class] = {}
+	end
+	bin[class][object] = tick()+expirationTime;
```
It might be better to pass this in as parameter.

Ignore this: I hadn't seen the new changes.

tberg-roblox reviewed (changes requested):
I really like this change. I left a few very minor comments and will pull this into the main branch when resolved.

Commit:
More adjustments from feedback

Should be final hopefully?
